java/statestore/src/main/java/sleeper/statestore/dynamodb/DynamoDBFileReferenceStore.java

```diff
@@ -257,6 +307,27 @@ public void atomicallyReplaceFileReferencesWithNewOne(
             double totalConsumed = consumedCapacity.stream().mapToDouble(ConsumedCapacity::getCapacityUnits).sum();
             LOGGER.debug("Removed {} file references and added 1 new file, capacity consumed = {}",
                     inputFiles.size(), totalConsumed);
+        } catch (TransactionCanceledException e) {
+            List<CancellationReason> reasons = e.getCancellationReasons();
+            for (int i = 0; i < inputFiles.size(); i++) {
+                if (isConditionCheckFailure(reasons.get(i))) {
+                    if (reasons.get(i).getItem() != null) {
+                        FileReference failedUpdate = fileReferenceFormat.getFileReferenceFromAttributeValues(reasons.get(i).getItem());
+                        if (!jobId.equals(failedUpdate.getJobId())) {
+                            throw new FileReferenceNotAssignedToJobException(failedUpdate, jobId);
+                        } else {
+                            throw new FileReferenceNotFoundException(failedUpdate);
+                        }
+                    } else {
+                        throw new FileReferenceNotFoundException(inputFiles.get(i), partitionId);
+                    }
+                }
+            }
+            CancellationReason writeReason = reasons.get(inputFiles.size());
+            if (isConditionCheckFailure(writeReason)) {
+                throw new FileReferenceAlreadyExistsException(fileReferenceFormat.getFileReferenceFromAttributeValues(writeReason.getItem()));
+            }
```

Comment:
This can be its own method to process the cancellation reasons, or split further. We can avoid discarding the original TransactionCanceledException.